### PR TITLE
33383 GraphQL: add new mutation 'deleteCustomer'

### LINKED ISSUE #33383: Add deleteCustomer mutation
*As a developer I want to be able to delete a customer via graphql so that I can build a button for customers to click on to delete their account when logged in to comply with the GDPR.*

### AC
* A mutation exists with the name 'deleteCustomer' which takes no arguments and returns a boolean and is capable of throwing an exception if anything went wrong.
* Calling the mutation will delete the customer identified by the customer token which must be sent along in the request header.

### PR DESCRIPTION
**Pull request by khrystynastolbova**
### Related Pull Requests

This PR added new  mutation 'deleteCustomer'

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33383

### Manual testing scenarios (*)

1. Generate customer token
2. Use customer tokens in the header of your GraphQL call
3. Run
```
mutation {
  deleteCustomer
}
```
Screenshot 2021-07-01 at 16 02 14

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by khrystynastolbova** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php]: Please move doctype declaration to a multiple row, as you already did above.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]:
```suggestion
    public function testDeleteCustomer(): void
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php:33]: Please add method's descriptions to **all** test cases.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]:
```suggestion
    public function testDeleteCustomerIfUserIsNotAuthorized(): void
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]:
```suggestion
    private function getMutation(): string
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]:
```suggestion
    public function testDeleteCustomerIfAccountIsLocked(): void
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]: I'm just wondering, is it possible that a locked customer to log in?

File: app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php:72]: I have a concern regarding the deprecated `\Magento\Framework\Registry` usage. Is there any alternative? Are we required to play with this configuration?

File: app/code/Magento/CustomerGraphQl/Model/Customer/DeleteCustomer.php
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/Model/Customer/DeleteCustomer.php]: Why not simply passing the `$context->getUserId()` instead as customer ID?
```suggestion
    public function execute(int $customerId): void
```

File: app/code/Magento/CustomerGraphQl/Model/Customer/DeleteCustomer.php
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/Model/Customer/DeleteCustomer.php]: Please update the description accordingly to the class's purpose.

**Review by eduard13 — CHANGES_REQUESTED** [review]
Thank you @khrystynastolbova for working on this one. Could you please review the questions and suggestions below?

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (khrystynastolbova) [reply dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]: You are right. We can't log in by the locked customer 
```
message": "The account sign-in was incorrect or your account is disabled temporarily. Please wait and try again later.",
```
This test was successful because the customer is used from the registry during authentication
`$this->customerRegistry->retrieve($customerId)` and there customer is saved as unlocked. 

This test failed if I use the `Magento/Customer/_files/locked_customer.php` fixture instead of 
```
Magento/Customer/_files/customer.php
 $this->lockCustomer->execute(1);
```
I will fix it. 
Thanks!

File: app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php
Comment (khrystynastolbova) [reply app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php:72]: The check is located in `magento/lib/internal/Magento/Framework/Model/ActionValidator/RemoveAction.php`
```
 public function isAllowed(AbstractModel $model)
    {
        $isAllowed = true;

        if ($this->registry->registry('isSecureArea')) {
            $isAllowed = true;
        } elseif (in_array($this->getBaseClassName($model), $this->protectedModels)) {
            $isAllowed = false;
        }
        return $isAllowed;
    }
```
ProtectedModels include the customer model too. In the file magento/app/code/Magento/Customer/etc/di.xml
```
<type name="Magento\Framework\Model\ActionValidator\RemoveAction">
        <arguments>
            <argument name="protectedModels" xsi:type="array">
                <item name="customer" xsi:type="string">Magento\Customer\Model\Customer</item>
            </argument>
        </arguments>
    </type>
```

**Comment by khrystynastolbova** [comment]
Hi @eduard13. Thank you for your suggestions. I made changes except the deprecated \Magento\Framework\Registry usage.

**Comment by khrystynastolbova** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php
Comment (khrystynastolbova) [reply app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php:72]: To avoid deprecated `\Magento\Framework\Registry`, we can remove the Customer model from protected models in general or copy all protected models to graphQL/di.xml for all magento modules except the Customer one.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]: Please adjust the description

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]: Is there any reason to call the parent method?

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/DeleteCustomerTest.php]: Lets avoid long code lines
```suggestion
        $this->expectExceptionMessage(
            'The account sign-in was incorrect or your account is disabled temporarily. Please wait and try again later'
        );
```

File: app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php]:
```suggestion
        $this->registry = $registry;
```

File: app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/DeleteCustomer.php:12]: Please sort the imports in ascending order

**Review by eduard13 — CHANGES_REQUESTED** [review]
Could you please review the below suggestions as well?
Thank you.

**Comment by eduard13** [comment]
@magento run Functional Tests B2B

**Comment by khrystynastolbova** [comment]
Hi @eduard13. Thanks for your review! I have applied your suggestions.

**Review by eduard13 — APPROVED** [review]

**Comment by eduard13** [comment]
@magento run all tests

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-9205](https://jira.corp.magento.com/browse/ENGCOM-9205) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Magento Health Index

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Cross checked the mutation to delete the customer and yes did work, and this PR is value addition.
Thanks for the contribution and collaboration @khrystynastolbova and @eduard13 
But there is a merge conflict with this PR and request you to kindly resolve it in order to test and progress further.

FYI: Since this customer deletion feature is not existing in UI, forwarded the query to PO upon it.

**Comment by engcom-Charlie** [comment]
Hi @engcom-Alfa,

I have updated the branch with 2.4-develop so moving it further for testing.

Thank you!

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
@magento run Static Tests, Magento Health Index, Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by engcom-Alfa** [comment]
@magento run Static Tests, Magento Health Index, Functional Tests CE

**Comment by engcom-Alfa** [comment]
@magento run Static Tests, Magento Health Index

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have Magento latest installed
2. Make sure to have at least one working customer login.


**Manual testing scenario:**

1. Get the customer token thru GraphQL request.

2. GraphQl request with below mutation, and validate the response
`mutation {
  deleteCustomer
}`


**Before:** :heavy_multiplication_x: We had no GraphQL mutation at all to delete the customer.
image


**After:** :heavy_check_mark: Customer with given token in the GraphQL request header gets deleted successfully
image


**Additional Testing scenarios:**
1. Deleted the customer thru above GraphQL mutation and tried to re-add the same customer from front end _Create an account_ option.
2. Kept the customer UI account with active login and deleted the account, validated the logged browser session get wiped put automatically.

There are no any additional testing is required as part of regression since there is an automation cases already covered.

**Comment by engcom-Alfa** [comment]
Since there are a couple of auto jobs are getting failed, and so moving the ticket to _Extended Testing_.

**Note:** Before moving the ticket to _Merge In Progress_, we should have a devdoc updates request on this case!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Integration Tests, Magento Health Index

**Comment by engcom-Charlie** [comment]
Raised Magento Health Index approval [JIRA](https://jira.corp.magento.com/browse/MC-43379) ticket. Waiting for approval now.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Magento Health Index , Functional Tests B2B, Functional Tests EE

**Comment by engcom-Charlie** [comment]
@cpartica can you please provide [JIRA](https://jira.corp.adobe.com/browse/AC-6413) approval for Magento health Index. 

Once we get the approval we can move this PR in Merge In Progress.

Thank you!

**Pull request merged**
